---
title: "R Notebook"
output: html_notebook
---

This notebook is a first draft to do the ^rimary analyze of the GInestier's dataset on subopulation breast cells.

Data were download at : https://www.ebi.ac.uk/arrayexpress/experiments/E-MTAB-4145/

From M&M : 
"Microarray processing and data analysis were performed at the ProfileXpert core facility (Bron, France). Total RNA (100 ng) was amplified and biotin-labeled with a GeneChip 3' IVT Express Kit, according to procedures from Affymetrix. Microarray analyses were performed with high-density oligonucleotide arrays (Human Genome U133 Plus 2 or Mouse Genome 430 2.0 Array, Affymetrix). 15 ??g of biotinylated cRNA was fragmented, and chip hybridization was performed according to the Affymetrix protocol (http://www.affymetrix.com/). Arrays were washed and stained with streptavidin-phycoerythrin (Invitrogen) in a Fluidics Station 450 (Affymetrix) according to the manufacturer's instructions. The arrays were scanned with a confocal laser (Genechip scanner 3000, Affymetrix). CEL files were generated with Affymetrix GeneChip Command Console (AGCC) software 3.0. The complete set of CEL files is available in the GEO database under accession number GSE56031 and the ArrayExpress database under accession number E-MTAB-4145. The obtained data were normalized with Affymetrix Expression Console software with the Robust Multiarray Average (RMA) statistical algorithm. Probesets were median-centered with Partek Genomic Suite software 6.6 (Partek)."


"These subpopulations were freshly isolated from mammary tissue, originating from reduction mammoplasties and flow sorted using four markers (EpCAM, CD10, CD49f, ALDH). Three subpopulations enriched in mammary stem cells (MaSCs) are designed as MaSC1, 2, 3; the luminal progenitor are designed as LP, and the mature luminal cells as mL1 and mL2."

Training : https://rawgit.com/bioinformatics-core-shared-training/microarray-analysis/master/affymetrix.nb.html
https://www.bioconductor.org/packages/devel/workflows/vignettes/maEndToEnd/inst/doc/MA-Workflow.html

```{r}
BiocManager::install("affy")
BiocManager::install("ArrayExpress")
BiocManager::install("pd.hugene.1.0.st.v1")
BiocManager::install("hugene10sttranscriptcluster.db")
BiocManager::install("oligo")
BiocManager::install("arrayQualityMetrics")
BiocManager::install("geneplotter")

#General Bioconductor packages
    library(Biobase)
    library(oligoClasses)

#Annotation and data import packages
    library(ArrayExpress)
    library(pd.hugene.1.0.st.v1)
    library(hugene10sttranscriptcluster.db)

#Quality control and pre-processing packages
    library(oligo)
    library(arrayQualityMetrics)
     
#Analysis and statistics packages
    library(limma)
    library(topGO)
    library(ReactomePA)
    library(clusterProfiler)
     
#Plotting and color options packages
    library(ggplot2)
    library(geneplotter)
    library(RColorBrewer)
    library(pheatmap)
     
#Formatting/documentation packages
   #library(rmarkdown)
   #library(BiocStyle)
    library(dplyr)
    library(tidyr)
```

# 1. Get data

The first step of the analysis is to download the raw data CEL files. These files are produced by the array scanner software and contain the measured probe intensities. The data we use have been deposited at ArrayExpress and have the accession code E-MTAB-2967.

Each ArrayExpress data set has a landing page summarizing the data set, and we use the getAEfunction from the ArrayExpress Bioconductor package to obtain the ftp links to the raw data files (Data from Palmieri et. al. on ArrayEpress).

With the code below, we download the raw data (also including annotation data) from ArrayExpress 
(2) by using the getAE-function. The data are saved in the raw_data_dir created above. The names of the downloaded files are returned as a list.


We will now have a closer look at the data we downloaded from ArrayExpress

```{r}
raw_data_dir <- "."
  
anno_AE <- getAE("E-MTAB-4145", path = raw_data_dir, type = "raw")

```


# 2. Import of annotation data and microarray expression data as "ExpressionSet"

We import the SDRF file with the read.delim function from the raw data folder in order to obtain the sample annotation.

The sample names are given in the column Array.Data.File of the SDRF data table and will be used as rownames for the SDRF file.

We turn the SDRF table into an AnnotatedDataFrame from the Biobase package that we will need later to create an ExpressionSet for our data (3).

```{r}
sdrf_location <- file.path(raw_data_dir, "E-MTAB-4145.sdrf.txt")
SDRF <- read.delim(sdrf_location)

rownames(SDRF) <- SDRF$Array.Data.File
SDRF <- AnnotatedDataFrame(SDRF)
```

We now create the Expression Set object raw_data, which contains array data, pheno data (from the SDRF file) as well as the information of the chip annotation package used.

The analysis of Affymetrix arrays starts with CEL files. These are the result of the processing of the raw image files using the Affymetrix software and contain estimated probe intensity values. Each CEL file additionally contains some metadata, such as a chip identifier.

We use the function read.celfiles from the oligo package (4) to import the files:

```{r}
raw_data <- oligo::read.celfiles(filenames = file.path(raw_data_dir, 
                                                SDRF$Array.Data.File),
                                    verbose = FALSE, phenoData = SDRF)
stopifnot(validObject(raw_data))
```
```{r}
exprs(raw_data) %>% head

exprs(raw_data)%>% density %>% plot
```

This automatically creates an ExpressionSet, fills the sections "array data" with the data from the CEL files and uses the correct chip annotation package, in this case pd.hugene.1.0.st.v1 (the chip-type is also stored in the .CEL files).

Furthermore, we specified our AnnotatedDataFrame "SDRF" created earlier from the SDRF file as phenoData. Thus, we had to make sure to import the CEL files in the order that corresponds to the SDRF table - to enforce this, we used the column Array.Data.File of the SDRF table as the filenames argument.

Finally, we checked whether the object created is valid (e.g. sample names match between the different tables).

We now have a first look on the raw data.

The pData function of the Biobase package directly accesses the phenoData in the ExpressionSet raw_data. With the head() function, we can view the first six lines of the table. We have a look at the columns included and retain only those columns that are related to the experimental factors of interest.

```{r}
head(Biobase::pData(raw_data))
```

The columns of interest for us are the following:

    identifiers of the individuals, i.e. columns "Source.Name", "Characteristics.individual.", "Array.Data.File"
    cell type, i.e. "Characteristics.cell.type.", "Characteristics.phenotype."

We now subselect the corresponding columns:
```{r}
Biobase::pData(raw_data) <- Biobase::pData(raw_data)[, c("Source.Name",
                                     "Characteristics.individual.", 
                                     "Array.Data.File",
                                     "Factor.Value.cell.type.",
                                     "Factor.Value.individual.")]
```




#  3. Quality control of the raw data

The first step after the initial data import is the quality control of the data. Here we check for outliers and try to see whether the data clusters as expected, e.g. by the experimental conditions. The expression intensity values are in the assayData sub-object "exprs" and can be accessed by the exprs(raw_data) function. The rows represent the microarray probes, i.e. the single DNA locations on the chip, while the columns represent one microarray, i.e. a sample of inflamed and non-inflamed tissue of every patient, respectively.

```{r}
Biobase::exprs(raw_data)[1:5, 1:5]
```

For quality control, we take the log2 of Biobase::exprs(raw_data), as expression data is commonly analyzed on a logarithmic scale.

We then perform a principal component analysis (PCA) and plot it (Figure 2). Every point in the plot represents one sample, with the colour indicating the mucosa type (inflamed vs non-inflamed) and the shape indicating the disease (UC or CD).

```{r}
exp_raw <- log2(Biobase::exprs(raw_data))
PCA_raw <- prcomp(t(exp_raw), scale. = FALSE)

percentVar <- round(100*PCA_raw$sdev^2/sum(PCA_raw$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA_raw$x[,1], PC2 = PCA_raw$x[,2],
                    CellType = pData(raw_data)$Factor.Value.cell.type.,
                    Individu = pData(raw_data)$Factor.Value.individual.)

ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = CellType, colour = Individu), size = 2) +
  ggtitle("PCA plot of the log-transformed raw expression data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5))+
  coord_fixed(ratio = sd_ratio) 
```


The PCA plot (Figure 2, performed on the log-intensity scale) of the raw data shows that the first principal component differentiates between the diseases. This means that the disease type is a major driver of gene expression differences. This might hinder our analysis, as we want to analyze the differential expression between inflamed and non-inflamed tissues, independently of the disease a person suffers from.

We also represent the probe intensities via a boxplot graph with one box per individual microarray. (Figure 3). Note that the oligo::boxplot function, i.e. the boxplot function of the oligo package, can take expression sets as argument. It accesses the expression data and performs a log2-transformation by default. We therefore can use raw_data as argument here.
```{r}
oligo::boxplot(raw_data, target = "core", 
               main = "Boxplot of log2-intensitites for the raw data")
```


When looking at the boxplot (Figure 3), we see that the intensity distributions of the individual arrays are quite different, indicating the need for an appropriate normalization, which we will discuss next.

Until now, we have only performed a very basic quality control; more elaborate quality control plots are available in the package arrayQualityMetrics (5). The package produces an html report, containing the quality control plots together with a description of their aims and an identification of possible outliers. We do not discuss this tool in detail here, but simply provide the code below, which creates a report for our raw data.
```{r}
arrayQualityMetrics(expressionset = raw_data,
    outdir = ".",
    force = TRUE, do.logtransform = TRUE,
    intgroup = c("Factor.Value.cell.type.", "Factor.Value.individual."))
```


# 4. Relative Log Expression data quality analysis

Before calibrating and evaluating the data, we want to perform another quality control procedure, namely Relative Log Expression (RLE), as described in the article by Gandolfo et al (9). To this end, we first perform an RMA without prior normalization:

```{r}
palmieri_eset <- oligo::rma(raw_data, target = "core", normalize = FALSE)
```

We finally reshape the data into a format in which we can use to create a boxplot for each array, as before:
```{r}
row_medians_assayData <- 
  Biobase::rowMedians(as.matrix(Biobase::exprs(palmieri_eset)))

RLE_data <- sweep(Biobase::exprs(palmieri_eset), 1, row_medians_assayData)

RLE_data <- as.data.frame(RLE_data)
RLE_data_gathered <- 
  tidyr::gather(RLE_data, patient_array, log2_expression_deviation)

ggplot2::ggplot(RLE_data_gathered, aes(patient_array,
                                       log2_expression_deviation)) + 
  geom_boxplot(outlier.shape = NA) + 
  ylim(c(-2, 2)) + 
  theme(axis.text.x = element_text(colour = "aquamarine4", 
                                  angle = 60, size = 6.5, hjust = 1 ,
                                  face = "bold"))
```

15 - 18 outlier s???



# 5. RMA calibration of the data
Now, we can apply the full RMA algorithm to our data in order to background-correct, normalize and summarize:

```{r}
bgData1 <- backgroundCorrect(raw_data)

palmieri_eset_norm <- oligo::rma(raw_data, target = "core")
```

```{r}
par(mfrow = c(1,2))

oligo::boxplot(raw_data, target = "core", 
               main = "log2-intensitites for the raw data")
oligo::boxplot(palmieri_eset_norm, target = "core", 
               main = "log2-intensitites for the norm data")
```


The parameter target defines the degree of summarization, the default option of which is "core", using transcript clusters containing "safely" annotated genes. For summarization on the exon level (not recommended for Gene arrays), one can use "probeset" as the target option. Although other methods for background correction and normalization exist, RMA is usually a good default choice. RMA shares information across arrays and uses the versatile quantile normalization method that will make the array intensity distributions match. However, it is preferable to apply it only after outliers have been removed. The quantile normalization algorithm used by RMA works by replacing values by the average of identically ranked (within a single chip) values across arrays. A more detailed description can be found on the Wikipedia page.

An alternative to quantile normalization would be the vsn algorithm,that performs background correction and normalization by robustly shifting and scaling intensity values within arrays before log-transforming them. This is less "severe" than quantile normalization (10).

"Factor.Value.cell.type.", "Factor.Value.individual."

```{r}
exp_palmieri <- Biobase::exprs(palmieri_eset_norm)
PCA <- prcomp(t(exp_palmieri), scale = FALSE)

percentVar <- round(100*PCA$sdev^2/sum(PCA$sdev^2),1)
sd_ratio <- sqrt(percentVar[2] / percentVar[1])

dataGG <- data.frame(PC1 = PCA$x[,1], PC2 = PCA$x[,2],
                    Disease = 
                     Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    Phenotype = 
                     Biobase::pData(palmieri_eset_norm)$Factor.Value.individual.)


ggplot(dataGG, aes(PC1, PC2)) +
      geom_point(aes(shape = Disease, colour = Phenotype)) +
  ggtitle("PCA plot of the calibrated, summarized data") +
  xlab(paste0("PC1, VarExp: ", percentVar[1], "%")) +
  ylab(paste0("PC2, VarExp: ", percentVar[2], "%")) +
  theme(plot.title = element_text(hjust = 0.5)) +
  coord_fixed(ratio = sd_ratio) 
```

```{r}
library(stringr)
phenotype_names <- pData(palmieri_eset_norm)$Factor.Value.cell.type.

disease_names <- pData(palmieri_eset_norm)$Factor.Value.individual.

annotation_for_heatmap <- 
  data.frame(Phenotype = phenotype_names,  Disease = disease_names)

row.names(annotation_for_heatmap) <- row.names(pData(palmieri_eset_norm))


dists <- as.matrix(dist(t(exp_palmieri), method = "manhattan"))

rownames(dists) <- row.names(pData(palmieri_eset_norm))
hmcol <- rev(colorRampPalette(RColorBrewer::brewer.pal(9, "YlOrRd"))(255))
colnames(dists) <- NULL
diag(dists) <- NA

ann_colors <- list(
  Phenotype = c(non_infl. = "chartreuse4", infl. = "burlywood3"),
  Disease = c(CD = "blue4", UC = "cadetblue2")
                   )
pheatmap(dists, col = (hmcol), 
         annotation_row = annotation_for_heatmap,
         legend = TRUE, 
         treeheight_row = 0,
         legend_breaks = c(min(dists, na.rm = TRUE), 
                         max(dists, na.rm = TRUE)), 
         legend_labels = (c("small distance", "large distance")),
         main = "Clustering heatmap for the calibrated samples")
```


# Present / Abs

```{r}
## Compute the A/M/P calls
ap <- paCalls(raw_data@experimentData, method="MAS5")
is(ap)


## Retrieve the A/M/P values from ap (we don't need the other attributes).
## These values are accessible via the same function exprs() used above
## to get expression values from the affy object.
ap <- exprs(ap)

## Compute a cross-table indicating the number of A, M and P calls for
## each sample.
ap.table <- apply(ap,2,table)
print(ap.table)


## Compute the relative frequencies of A, M and P calls per sample
ap.freq.table <- ap.table / apply(ap.table, 2, sum)
print(ap.freq.table)

## Draw a barplot indicating the relative frequencies of A, M and P calls per sample.
barplot(ap.table/nrow(ap),las=2, cex.names=0.55, main="% of A/M/P calls")


## Let us add a legend besides the barplots
barplot(ap.table/nrow(ap),las=2, cex.names=0.55, main="% of A/M/P calls", legend=rownames(ap.table), xlim=c(1,18))
```


```{r}
pres_abs <- paCalls(raw_data) 

pres_abs %>% head
```







# 6. Filtering based on intensity

We now filter out lowly expressed genes. Microarray data commonly show a large numberof probes in the background intensity range. These probes also do not change much across arrays. Hence they combine a low variance with a low intensity. Thus, they could end up being detected as differentially expressed although they are barely above the "detection" limit and are not very informative in general.

We will perform a "soft" intensity based filtering here, since this is recommended by the limma (11,12) user guide (a package we will use below for the differential expression analysis).

However, note that a variance based filter might exclude a similar set of probes in practice. For intensity-based filtering, we calculate the row-wise medians from the expression data, as they represent the transcript medians, and assign them to palmieri_medians. From this we create a histogram:
```{r}
palmieri_medians <- rowMedians(Biobase::exprs(palmieri_eset_norm))

hist_res <- hist(palmieri_medians, 100, col = "cornsilk1", freq = FALSE, 
            main = "Histogram of the median intensities", 
            border = "antiquewhite4",
            xlab = "Median intensities")
```

In the histogram of the gene-wise medians (Figure 8), we can clearly see an enrichment of low medians on the left hand side. These represent the genes we want to filter. In order to infer a cutoff from the data, we inspect the histogram: We visually set a cutoff line man_threshold to the left of the histogram peak in order not to exclude too many genes. In our example, we choose a threshold of 4. We plot the same histogram as before and add the threshold line with the abline() function:
```{r}
man_threshold <- median(exprs(palmieri_eset_norm))

hist_res <- hist(palmieri_medians, 100, col = "cornsilk", freq = FALSE, 
            main = "Histogram of the median intensities",
            border = "antiquewhite4",
            xlab = "Median intensities")

abline(v = man_threshold, col = "coral4", lwd = 2)
```

ICI ON NE FILTRE PAS JE VEUX HORMAD1 ET CT83


```{r}
palmieri_manfiltered <- palmieri_eset_norm
```



# 7. Annotation
```{r}
anno_palmieri <- AnnotationDbi::select(hugene10sttranscriptcluster.db,
                                  keys = (featureNames(palmieri_manfiltered)),
                                  columns = c("SYMBOL", "GENENAME"),
                                  keytype = "PROBEID")

anno_palmieri <- subset(anno_palmieri, !is.na(SYMBOL))
```
Removing multiple mappings

Many transcript-cluster identifiers will map to multiple gene symbols, i.e. they can't be unambigously assigned.

We compute a summary table in the code below to see how many there are:


```{r}
anno_grouped <- group_by(anno_palmieri, PROBEID)
anno_summarized <- 
  dplyr::summarize(anno_grouped, no_of_matches = n_distinct(SYMBOL))

head(anno_summarized)


anno_filtered <- filter(anno_summarized, no_of_matches > 1)

head(anno_filtered)


probe_stats <- anno_filtered 

nrow(probe_stats)
```

First, we grouped anno_palmieri by their PROBEID; that way, the subsequent operations are not carried through for each single row, but for each group, i.e. each PROBEID. We then summarized the groups and indicate the number of different genes assigned to a transcript cluster in the column no_of_matches. Finally, we filtered for PROBEIDs with multiple matches, i.e. no_of_matches > 1.

With dim(probe_stats), we could see how many probes have been mapped to multiple genes.

We have close to 2000 transcript clusters that map to multiple gene symbols. It is difficult to decide which mapping is "correct". Therefore, we exclude these transcript clusters.

We want to remove those probe IDs that match the ones in probe_stats, as those are the probes with multiple mappings. We assign these IDs to the variable ids_to_exclude. Then, we generate palmieri_final, an expression set without the ids_to_exclude.
```{r}
ids_to_exlude <- (featureNames(palmieri_manfiltered) %in% probe_stats$PROBEID)

table(ids_to_exlude)
palmieri_final <- subset(palmieri_manfiltered, !ids_to_exlude)

validObject(palmieri_final)

head(anno_palmieri)
```

Recall that fData enables us to access the feature data of an expression set. Until now, no feature data whatsoever is stored in the fData(palmieri_final). Only the row names are the row names of the assay data by default, which are the PROBEIDs of the transcripts.

Therefore, we generate a column PROBEID in fData(palmieri_final) and assign the row names of fData(palmieri_final) to it:


Then, we left-join fData(palmieri_final)with anno_palmieri, which already contains the columns "SYMBOL" and "GENENAME". A left-join keeps the rows and columns of the first argument and adds the corresponding column entries of the second argument:



```{r}
fData(palmieri_final)$PROBEID <- rownames(fData(palmieri_final))

fData(palmieri_final) <- left_join(fData(palmieri_final), anno_palmieri)
# restore rownames after left_join
rownames(fData(palmieri_final)) <- fData(palmieri_final)$PROBEID 
    
validObject(palmieri_final)

```

# Rank
```{r}
exp_palmieri <- Biobase::exprs(palmieri_final)
med <- rowMedians(Biobase::exprs(palmieri_final))


st_exp_palmieri <- stack(exp_palmieri)
head(st_exp_palmieri$value %>% sort)
```


```{r}
dt_sort <- data.frame(number = order(sort(st_exp_palmieri$value)),
                      value = sort(st_exp_palmieri$value))
# Combien de probes sous le thr ? 
sum(dt_sort$value <= quantile(dt_sort$value)[4])/20
# Combien de probes au dessus ? 
sum(dt_sort$value >= quantile(dt_sort$value)[4])/20

ggplot(dt_sort[sample(nrow(dt_sort), 1000), ], aes( x= number, y = value))+
  geom_point()+
  geom_hline(yintercept = quantile(dt_sort$value)[2])+
  geom_hline(yintercept = quantile(dt_sort$value)[3], color = "red")+
  geom_hline(yintercept = quantile(dt_sort$value)[4])

```

```{r}
dt_sort <- data.frame(number = order(sort(med)),
                      value = sort(med))

# Combien de probes sous le thr ? 
sum(dt_sort$value <= quantile(dt_sort$value)[4])
# Combien de probes au dessus ? 
sum(dt_sort$value >= quantile(dt_sort$value)[4])

ggplot(dt_sort[sample(nrow(dt_sort), 1000), ], aes( x= number, y = value))+
  geom_point()+
  geom_hline(yintercept = quantile(dt_sort$value)[2])+
  geom_hline(yintercept = quantile(dt_sort$value)[3], color = "red")+
  geom_hline(yintercept = quantile(dt_sort$value)[4])
```
```{r}
thr <- quantile(dt_sort$value)[4]
thr_l <- quantile(dt_sort$value)[2]

```


# 8 BOXPLOT

## HORMAD1 CT
```{r}
exp_palmieri <- Biobase::exprs(palmieri_final)

A = "HORMAD1"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g1 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l )+
  geom_hline(yintercept = man_threshold, col = "red4")+
  annotate(geom="text", x=4, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                  data = dataGG)))[9], 3)),
           color="black")+
  
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
    scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+
  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")


A = "CT83"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g2 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
  geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+
  
  annotate(geom="text", x=4, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                  data = dataGG)))[9], 3)),
           color="black")+
  
  ggtitle("boxplot of the calibrated  data") +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
    scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+
  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")

unlist(summary(aov(formula = y~x, data = dataGG)))[9]

library(gridExtra)
grid.arrange(g1, g2, ncol = 2)

```

## Other CT genes
```{r}
exp_palmieri <- Biobase::exprs(palmieri_final)

A = "DMRTC2"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g1 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l )+
  geom_hline(yintercept = man_threshold, col = "red4")+
  annotate(geom="text", x=4, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                  data = dataGG)))[9], 3)),
           color="black")+
  
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
    scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+
  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")


A = "LRGUK"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g2 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
  geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+
  
  annotate(geom="text", x=4, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                  data = dataGG)))[9], 3)),
           color="black")+
  
  ggtitle("boxplot of the calibrated  data") +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
    scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+
  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")

unlist(summary(aov(formula = y~x, data = dataGG)))[9]

library(gridExtra)
grid.arrange(g1, g2, ncol = 2)
```
```{r}
exp_palmieri <- Biobase::exprs(palmieri_final)

A = "TEX14"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g1 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l )+
  geom_hline(yintercept = man_threshold, col = "red4")+
  annotate(geom="text", x=4, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                  data = dataGG)))[9], 3)),
           color="black")+
  
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
    scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+
  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")


A = "TDRD1"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g2 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
  geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+
  
  annotate(geom="text", x=4, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                  data = dataGG)))[9], 3)),
           color="black")+
  
  ggtitle("boxplot of the calibrated  data") +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
    scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+
  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")

unlist(summary(aov(formula = y~x, data = dataGG)))[9]

library(gridExtra)
grid.arrange(g1, g2, ncol = 2)
```



## Selected larkers

Pour les marqueurs, tu peux utiliser ITGA6 (CD49f) ou n'importe quelle marqueur d'EMT (ZEB1, VIM) pour les MaSC. Pour toutes les autres population tu peux utiliser ESA (CD326) ou encore le recepteur aux oestrogènes (ESR1).

Les kératines sont aussi des bons marqueurs (KRT6A, KRT8, KRT14).


```{r}
grid.arrange(g4, g5,g3, ncol = 3)

```

```{r}
grid.arrange( g6,g7, ncol = 2)

```




## MASC
```{r}
exp_palmieri <- Biobase::exprs(palmieri_final)

A = "ZEB1"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g6 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+  annotate(geom="text", x=3, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                 data = dataGG)))[9], 5)),
          color="black")+

  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
  scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+

  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")
g6
```

```{r}
A = "VIM"
A = "MSRB3"
id_A <- grep(A, fData(palmieri_final)$SYMBOL)

fData(palmieri_final)$SYMBOL[id_A]

dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g7 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+ annotate(geom="text", x=3, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                                 data = dataGG)))[9], 10)),
          color="black")+

  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
  scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+

  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")
g7
```

# Mature
```{r}
A = "EPCAM"  # CD326
fData(palmieri_final)$SYMBOL[id_A]

id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[2],] )
g5 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+ annotate(geom="text", x=3, y=6, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, 
                                                                              data = dataGG)))[9], 10)),
          color="black")+

  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
  scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+
  labs(title="EpCAM (ESA, CD326)", x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")
g5
```

```{r}
A = "ESR1" 
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[2],] )
g4 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+ annotate(geom="text", x=3, y=11.5, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, data = dataGG)))[9], 10)), color="black")+

  ggtitle("boxplot of the calibrated  data") +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
      scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+

  labs(title=paste(A), x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")
g4
```


```{r}
A = "ITGA6" # CD49f
id_A <- grep(A, fData(palmieri_final)$SYMBOL)
dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
g3 = ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+ annotate(geom="text", x=3, y=6, label=paste("AOV:", round(unlist(summary(aov(formula = y~x, data = dataGG)))[9], 5)), color="black")+

  ggtitle("boxplot of the calibrated  data") +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
      scale_y_continuous(limits = c(4,12))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+

  labs(title="CD49f (ITGA6)", x="Subtype",y="Expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")
g3
```


```{r}
A = "VIM" 
id_A <- grep(A, fData(palmieri_final)$SYMBOL)

 fData(palmieri_final)$SYMBOL[id_A]

dataGG <- data.frame(x =Biobase::pData(palmieri_eset_norm)$Factor.Value.cell.type.,
                    y =exp_palmieri[id_A[1],] )
ggplot(dataGG, aes(x, y, colour = x, fill = x)) +
  geom_boxplot(outlier.colour = "white", color = "black")+
      geom_jitter( width = 0.1, size = 3, shape = 21, color = "black") +
  geom_hline(yintercept = thr )+
  geom_hline(yintercept = thr_l)+
  geom_hline(yintercept = man_threshold, col = "red4")+ annotate(geom="text", x=5, y=11.5, 
                                                                 label=paste("AOV:",    round(unlist(summary(aov(formula = y~x, data = dataGG)))[9], 5)), color="black")+

  ggtitle("boxplot of the calibrated  data") +
  theme(plot.title = element_text(hjust = 0.5))  + 
  scale_fill_manual(values=c("cyan4","orange4","goldenrod3", "orange3", "green3", "green4"))+
  scale_color_manual(values=c("cyan3","orange2","goldenrod2", "orange1", "green2", "green3"))+
      scale_y_continuous(limits = c(4,12.5))+
  scale_x_discrete(limits= c("mammary stem cell 1", "mammary stem cell 2", "mammary stem cell 3",
                     "luminal progenitor cell", "mature luminal cell 1",   "mature luminal cell 2" ))+

  labs(title=paste(A), x="Subtype",y="expression")+                         
  theme_classic()+                                                     # thème blanc
  theme(plot.title = element_text(size = 12, face = "bold",hjust=0.5),    #titre en gras, centré
              text=element_text(),
              axis.title = element_text(face="bold", size=10),        #titre des axes en gras
              axis.text.x=element_text(angle=45, hjust=1,colour="black", size = 8),
              axis.text.y=element_text(colour="black", size = 8),
              legend.position = "none")
```





# 8. Linear models

In order to analyse which genes are differentially expressed between inflamed and non-inflamed tissue, we have to fit a linear model to our expression data. Linear models are the "workhorse" for the analysis of experimental data. They can be used to analyse almost arbitrarily complex designs, however they also take a while to learn and understand and a thorough description is beyond the scope of this workflow.

Mike Love's and Rafael Irizzary's "Biomedical Data Science" (14) is a very good resource, especially the section on interactions and contrasts. It might also be helpful to learn some linear algebra to better understand the concepts here. The Khan Academy offers helpful (and free) online courses.

















```{r}
Sample_and_data_relationship <- read.delim("~/Desktop/These_Marthe/1_Bioinfo/200505_MicroarrayGinestier_E-MTAB-4145/Sample_and_data_relationship.txt")

head(Sample_and_data_relationship)
```



```{r}
library(affy)
targetsFile <- Sample_and_data_relationship$Array.Data.File

# the row.names of your phenoData object should be identical to what you get from list.celfiles()
rownames(Sample_and_data_relationship) <- targetsFile
targetsFile
```


The function to import the data is ReadAffy from the affy package.

```{r}
raw <-ReadAffy(celfile.path = "./E-MTAB-4145.raw.1", 
               filenames = targetsFile,
               phenoData = Sample_and_data_relationship)
raw
```
=> Info : 20 samples, 32 000 probes





